File: library/src/main/java/io/ashdavies/rx/rxfirebase/RxFirebaseDatabase.java
package io.ashdavies.rx.rxfirebase;

import android.support.annotation.CheckResult;
import android.support.annotation.Nullable;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Logger;
import com.google.firebase.database.Query;
import io.ashdavies.rx.rxtasks.RxTasks;
import io.reactivex.BackpressureStrategy;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import java.util.Map;

@SuppressWarnings("WeakerAccess")
public final class RxFirebaseDatabase {

  private final Query query;

  private RxFirebaseDatabase(Query query) {
    this.query = query;
  }

  public static RxFirebaseDatabase getInstance() {
    return getInstance(FirebaseDatabase.getInstance().getReference());
  }

  public static RxFirebaseDatabase getInstance(String path, Object... args) {
    return getInstance(FirebaseDatabase.getInstance().getReference(String.format(path, args)));
  }

  public static RxFirebaseDatabase getInstance(Query query) {
    return new RxFirebaseDatabase(query);
  }

  @CheckResult
  public RxFirebaseDatabase child(String child) {
    return getInstance(query.getRef().child(child));
  }

  @CheckResult
  public RxFirebaseDatabase orderByChild(String child) {
    return getInstance(query.orderByChild(child));
  }

  @CheckResult
  public RxFirebaseDatabase orderByKey() {
    return getInstance(query.orderByKey());
  }

  @CheckResult
  public RxFirebaseDatabase orderByPriority() {
    return getInstance(query.orderByPriority());
  }

  @CheckResult
  public RxFirebaseDatabase orderByValue() {
    return getInstance(query.orderByValue());
  }

  @CheckResult
  public RxFirebaseDatabase limitToFirst(int i) {
    return getInstance(query.limitToFirst(i));
  }

  @CheckResult
  public RxFirebaseDatabase limitToLast(int i) {
    return getInstance(query.limitToLast(i));
  }

  @CheckResult
  public Completable setPriority(Object object) {
    return RxTasks.completable(query.getRef().setPriority(object));
  }

  @CheckResult
  public Completable setValue(Object value) {
    return setValue(value, null);
  }

  @CheckResult
  public Completable setValue(Object value, @Nullable Object priority) {
    return Completable.create(new SetValueOnSubscribe(query.getRef(), value, priority));
  }

  @CheckResult
  public Completable updateChildren(Map<String, Object> map) {
    return RxTasks.completable(query.getRef().updateChildren(map));
  }

  @CheckResult
  public Completable removeValue() {
    return RxTasks.completable(query.getRef().removeValue());
  }

  @CheckResult
  public Flowable<DataSnapshot> onValueEvent() {
    return Flowable.create(new ValueEventOnSubscribe(query), BackpressureStrategy.BUFFER);
  }

  @CheckResult
  public Single<DataSnapshot> onSingleValueEvent() {
    return Single.create(new SingleValueEventOnSubscribe(query));
  }

  @CheckResult
  public Flowable<ChildEvent> onChildEvent() {
    return Flowable.create(new ChildEventOnSubscribe(query), BackpressureStrategy.BUFFER);
  }

  @CheckResult
  public Flowable<ChildEvent> onChildEvent(ChildEvent.Type childEventType) {
    return onChildEvent().filter(new ChildEventTypePredicate(childEventType));
  }

  @CheckResult()
  public <T> Flowable<T> onChildEventValue(ChildEvent.Type childEventType, Class<T> kls) {
    return onChildEvent(childEventType).compose(new ChildEventClassTransformer<>(kls));
  }

  @SuppressWarnings("UnusedReturnValue")
  public RxFirebaseDatabase purgeOutstandingWrites() {
    query.getRef().getDatabase().purgeOutstandingWrites();
    return this;
  }

  @SuppressWarnings("UnusedReturnValue")
  public RxFirebaseDatabase goOnline() {
    query.getRef().getDatabase().goOnline();
    return this;
  }

  @SuppressWarnings("UnusedReturnValue")
  public RxFirebaseDatabase goOffline() {
    query.getRef().getDatabase().goOffline();
    return this;
  }

  @SuppressWarnings("UnusedReturnValue")
  public RxFirebaseDatabase setLogLevel(Logger.Level level) {
    query.getRef().getDatabase().setLogLevel(level);
    return this;
  }
}

File: src/main/kotlin/io/ashdavies/rx/rxfirebase/RxFirebaseDatabase.kt
package io.ashdavies.rx.rxfirebase

import androidx.annotation.CheckResult
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.Logger
import com.google.firebase.database.Query
import io.ashdavies.rx.extensions.onChildEvent
import io.ashdavies.rx.extensions.onSingleValueEvent
import io.ashdavies.rx.extensions.onValueEvent
import io.ashdavies.rx.rxtasks.toCompletable
import io.reactivex.BackpressureStrategy
import io.reactivex.BackpressureStrategy.BUFFER
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import java.lang.String.format

@Deprecated("Use extension functions and/or RxTasks instead")
data class RxFirebaseDatabase(private val query: Query) {

  @CheckResult
  fun child(child: String): RxFirebaseDatabase = RxFirebaseDatabase(query.ref.child(child))

  @CheckResult
  fun orderByChild(child: String): RxFirebaseDatabase = RxFirebaseDatabase(query.orderByChild(child))

  @CheckResult
  fun orderByKey(): RxFirebaseDatabase = RxFirebaseDatabase(query.orderByKey())

  @CheckResult
  fun orderByPriority(): RxFirebaseDatabase = RxFirebaseDatabase(query.orderByPriority())

  @CheckResult
  fun orderByValue(): RxFirebaseDatabase = RxFirebaseDatabase(query.orderByValue())

  @CheckResult
  fun limitToFirst(limit: Int): RxFirebaseDatabase = RxFirebaseDatabase(query.limitToFirst(limit))

  @CheckResult
  fun limitToLast(limit: Int): RxFirebaseDatabase = RxFirebaseDatabase(query.limitToLast(limit))

  @CheckResult
  fun setPriority(priority: Any): Completable = query
      .ref
      .setPriority(priority)
      .toCompletable()

  @CheckResult
  @JvmOverloads
  fun setValue(value: Any, priority: Any? = null): Completable = query
      .ref
      .setValue(value, priority)
      .toCompletable()

  @CheckResult
  fun updateChildren(map: Map<String, Any>): Completable = query
      .ref
      .updateChildren(map)
      .toCompletable()

  @CheckResult
  fun removeValue(): Completable = query
      .ref
      .removeValue()
      .toCompletable()

  @CheckResult
  @JvmOverloads
  fun onValueEvent(strategy: BackpressureStrategy = BUFFER): Flowable<DataSnapshot> = query
      .onValueEvent()
      .toFlowable(strategy)

  @CheckResult
  fun onSingleValueEvent(): Single<DataSnapshot> = query.onSingleValueEvent()

  @CheckResult
  @JvmOverloads
  fun onChildEvent(strategy: BackpressureStrategy = BUFFER): Flowable<ChildEvent> = query
      .onChildEvent()
      .toFlowable(strategy)

  @CheckResult
  @JvmOverloads
  fun onChildEvent(type: ChildEvent.Type, strategy: BackpressureStrategy = BUFFER): Flowable<ChildEvent> = query
      .onChildEvent(type)
      .toFlowable(strategy)

  @CheckResult
  @JvmOverloads
  fun <T> onChildEventValue(kls: Class<T>, strategy: BackpressureStrategy = BUFFER): Flowable<T> {
    return onChildEvent(strategy).flatMapMaybe(ChildEventMapper(kls))
  }

  @CheckResult
  @JvmOverloads
  fun <T> onChildEventValue(type: ChildEvent.Type, kls: Class<T>, strategy: BackpressureStrategy = BUFFER): Flowable<T> {
    return onChildEvent(type, strategy).flatMapMaybe(ChildEventMapper(kls))
  }

  fun purgeOutstandingWrites(): RxFirebaseDatabase = apply {
    query
        .ref
        .database
        .purgeOutstandingWrites()
  }

  fun goOnline(): RxFirebaseDatabase = apply {
    query
        .ref
        .database
        .goOnline()
  }

  fun goOffline(): RxFirebaseDatabase = apply {
    query
        .ref
        .database
        .goOffline()
  }

  fun setLogLevel(level: Logger.Level): RxFirebaseDatabase = apply {
    query
        .ref
        .database
        .setLogLevel(level)
  }

  companion object {

    private val reference: DatabaseReference
      get() = FirebaseDatabase
          .getInstance()
          .reference

    private fun getReference(path: String, vararg args: Any): DatabaseReference = FirebaseDatabase
        .getInstance()
        .getReference(format(path, *args))

    @JvmStatic
    fun getInstance() = RxFirebaseDatabase(reference)

    @JvmStatic
    fun getInstance(path: String, vararg args: Any) = RxFirebaseDatabase(getReference(format(path, *args)))
  }
}
